### DISCUSSION TITLE
ERG mode based on HR (HR Zone based Training)

### DISCUSSION THREAD

**Discussion by supreik** [dvmarinoff/Auuki#177] [Ideas]
Hi, 

i've found your app and it's great. Maybe it's possible to add an additiontiol trainer mode: HRerg

The logic for adjusting watts based on a target heart rate could be as follows:
- The user sets a target heart rate and an initial start with 40% FTP.
- The training starts in "HR Tracking" mode.
- If the heart rate exceeds the upper limit, the percentage of FTP is reduced by a certain value (x) every minute outside the desired range.
- If the heart rate falls below the lower limit, the percentage of FTP is increased by another value (y) every minute outside the desired range.
- The adjustments are made slowly to avoid abrupt changes.
- These adjustments continue until the training is completed and the target heart rate is reached within +/- 1%.

What do you think?

**Comment by dvmarinoff** [comment]
Hi and welcome to the project repo!

Heart Rate ERG mode, is a bit hard to make right, and currently I don’t have the time to start the whole R&D process on the topic.
But about 2 years ago another User here, contacted me about help extending their fork (copy) of the project with this particular feature. They have the feature working on their fork of the project, but it's unmaintained and is still using the code from 2 years ago, when this app was just starting.

You can try to test the feature over there to see if it works: https://vo2max-ai.vercel.app/
And here is the discussion from back then: https://github.com/dvmarinoff/Flux/issues/76

Their implementation uses a file called `ml.js`, I assume maybe this is a Machine Learning model, so it might work well, but the file is like 34 MB large and it’s a no go for this project.

**Comment by supreik** [comment]
Wow, that was quick response. I will give it a try. 

Is the fork available within github? Or the ml.js?

**Reply by dvmarinoff** [reply]
Yes, it's here: https://github.com/loopingleo/VO2max

**Comment by bobmorane06** [comment]
I tried auuki for the first time today, congratulations for a very neat app! 

I was looking exactly for that feature, setting a workout based on target HR rather than power. This thread is a year old, but I wonder if there was now a way to achieve HR-based ERG ?

**Reply by dvmarinoff** [reply]
The good news: I am getting very much interested in Heart Rate based training
The bad news: it's going to take some time to create this feature

Another obstacle is that there is now commonly accepted format for creating Heart Rate based workouts. The few platforms that support heart rate workouts have very different ideas. Will take time to solve it, but currently I am also interested.
